src/app/api/webhooks/multibaas/route.ts

```diff
@@ -38,33 +46,234 @@ export async function POST(request: NextRequest): Promise<Response> {
+
+    await handleVoteEvent(event);
+    await handleProposalEvent(event);
   }
 
   return new Response(null, { status: 200 });
 }
 
 function assertSignature(
-  payload: unknown,
+  payload: string,
   signature: string | null,
   timestamp: string | null,
-): payload is MultibassEvent[] {
+): MultibassEvent[] | false {
   if (!payload || !signature || !timestamp) {
     return false;
   }
 
   const hmac = createHmac('sha256', process.env.MULTIBAAS_WEBHOOK_SECRET!);
-  hmac.update(Buffer.from(JSON.stringify(payload)));
+  hmac.update(Buffer.from(payload));
   hmac.update(timestamp);
   const signature_ = hmac.digest().toString('hex');
 
   if (signature !== signature_) {
     console.log({
       payload,
-      json: JSON.stringify(payload),
+      json: payload,
       signature,
       signature_,
     });
     return false;
   }
-  return true;
+  return JSON.parse(payload);
+}
+
+async function handleVoteEvent(event: MultibassEvent['data']['event']): Promise<void> {
+  if (
+    !assertEvent(
+      ['ProposalVoted', 'ProposalVotedV2', 'ProposalVoteRevoked', 'ProposalVoteRevokedV2'] as const,
+      event.name,
+    )
+  ) {
+    return;
+  }
+
+  const proposal = decodeVoteEventLog(JSON.parse(event.rawFields));
+  if (!proposal || !proposal.proposalId) {
+    throw new Error('Couldnt update the votes');
+  }
+
+  const { totals } = await fetchProposalVoters(proposal.proposalId);
+  const { count } = await database
+    .insert(votesTable)
+    .values(
+      Object.entries(totals).map(([type, count]) => ({
+        type: type as VoteType,
+        count,
+        chainId: celoPublicClient.chain.id,
+        proposalId: proposal.proposalId,
+      })),
+    )
+    .onConflictDoUpdate({
+      set: { count: sql`excluded.count` },
+      target: [votesTable.proposalId, votesTable.type, votesTable.chainId],
+    });
+
+  console.log(`Inserted ${count} vote records for proposal ${proposal.proposalId}`);
+}
+
+async function handleProposalEvent(event: MultibassEvent['data']['event']): Promise<void> {
+  if (
+    !assertEvent(
+      [
+        'ProposalExecuted',
+        'ProposalApproved',
+        'ProposalExpired',
+        'ProposalDequeued',
+        'ProposalQueued',
+      ] as const,
+      event.name,
+    )
+  ) {
+    return;
+  }
+
+  const { topics, data } = JSON.parse(event.rawFields);
+  const {
+    args: { proposalId },
+  } = decodeEventLog({
+    abi: governanceABI,
+    topics,
+    data,
+    eventName: event.name,
+  });
+  if (!proposalId) {
+    throw new Error('Couldnt update the proposal');
+  }
+
+  const cached = CachedMetadata as ProposalMetadata[];
+  const proposalsMetadata = await fetchProposalsFromRepo(cached, false);
+  const [_proposer, _deposit, _timestampSec, _numTransactions, url, _networkWeight, _isApproved] =
+    await celoPublicClient.readContract({
+      address: Addresses.Governance,
+      abi: governanceABI,
+      functionName: 'getProposal',
+      args: [proposalId],
+    });
```

Comment:
seem to me that we would only need to fetch the url for ProposalQueued event. after that its already going to exist.

Reply:
Good catch